File: app/src/main/java/jp/juggler/subwaytooter/util/WorkerBase.java
package jp.juggler.subwaytooter.util;

public abstract  class WorkerBase extends Thread{

	private static final LogCategory log = new LogCategory( "WorkerBase" );

	public abstract void cancel();
	public abstract void run();
	
	public void notifyEx(){
		try{
			synchronized( this ){
				notify();
			}
		}catch(Throwable ex){
			log.trace(ex);
		}
	}
	
	
	public void waitEx( long ms ){
		try{
			synchronized( this ){
				wait( ms );
			}
		}catch(Throwable ex){
			log.trace(ex);
		}
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/util/WorkerBase.kt
package jp.juggler.subwaytooter.util

import jp.juggler.util.coroutine.AppDispatchers.withTimeoutSafe
import jp.juggler.util.coroutine.launchDefault
import kotlinx.coroutines.Job
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.channels.Channel

abstract class WorkerBase(
    private val waiter: Channel<Unit> = Channel(capacity = Channel.CONFLATED),
) {

    private val suspendJob: Job

    abstract fun cancel()

    abstract suspend fun run()

    suspend fun waitEx(ms: Long) = try {
        withTimeoutSafe(ms) { waiter.receive() }
    } catch (ignored: TimeoutCancellationException) {
        null
    }

    fun notifyEx() = waiter.trySend(Unit)

    val isAlive: Boolean
        get() = suspendJob.isActive

    init {
        suspendJob = launchDefault { run() }
    }
}
